itests/fevm_test.go

```diff
@@ -1299,3 +1303,46 @@ func TestEthGetTransactionCount(t *testing.T) {
 	require.NoError(t, err)
 	require.Zero(t, contractNonceAfterDestroy)
 }
+
+func TestMcopy(t *testing.T) {
+	// MCOPY introduced in nv24, start the test on nv23 to check the error, then upgrade at epoch 100
+	// and check that an MCOPY contract can be deployed and run.
+	nv24epoch := abi.ChainEpoch(100)
+	upgradeSchedule := kit.UpgradeSchedule(
+		stmgr.Upgrade{
+			Network: network.Version23,
+			Height:  -1,
+		},
+		stmgr.Upgrade{
+			Network:   network.Version24,
+			Height:    nv24epoch,
+			Migration: filcns.UpgradeActorsV15,
+		},
+	)
+
+	ctx, cancel, client := kit.SetupFEVMTest(t, upgradeSchedule)
+	defer cancel()
+
+	// TODO: below here ------------------------------------------------------------------------------
+
+	// try to deploy the contract before the upgrade, expect an error somewhere' in deploy or in call,
+	// if the error is in deploy we may need to implement DeployContractFromFilename here where we can
+	// assert an error
+	filenameActor := "contracts/mcopy/MCOPYTest.hex"
+	fromAddr, contractAddr := client.EVM().DeployContractFromFilename(ctx, filenameActor)
+	_, _, err := client.EVM().InvokeContractByFuncName(ctx, fromAddr, contractAddr, "optimizedCopy(string)", []byte("testdata"))
+	// We expect an error here due to the contract reverting or another issue.
+	require.Error(t, err)
+
+	// Also check for the specific error message
+	expectedErrMsg := "contract reverted (33)"
+	require.Contains(t, err.Error(), expectedErrMsg)
+
+	// wait for the upgrade
+	client.WaitTillChain(ctx, kit.HeightAtLeast(nv24epoch+5))
+
+	// should be able to deploy and call the contract now
+	fromAddr, contractAddr = client.EVM().DeployContractFromFilename(ctx, filenameActor)
+	_, _, err = client.EVM().InvokeContractByFuncName(ctx, fromAddr, contractAddr, "optimizedCopy(string)", []byte("testdata"))
```

Comment:
This isn't the correct format for the parameters, unless I'm mistaken. This needs to be encoded as a variable-length string. See https://docs.soliditylang.org/en/latest/abi-spec.html#use-of-dynamic-types.

In this case, we need (concatenated):

1. 32 byte word indicating the offset of `testdata` (in bytes). So, `int256(32).to_bigendian()`.
2. The length of the data `int256(len("testdata")).to_bigendian()`.
3. "testdata", padded with zeros on the right to the next multiple of 32 bytes.

I usually use tools like: https://abi.hashex.org/

That gives us

```
683f3ac5000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000087465737464617461000000000000000000000000000000000000000000000000
```

Which includes the function selector at the beginning, you'll want to remove that.

Reply:
thank you for explaining this multiple times!